### ISSUE TITLE
Option to disable alt-click move cursor to position behavior

### ISSUE THREAD

**Issue by matthias-ccri** [xtermjs/xterm.js#3145]
Hi all,

I've been running into an issue that I don't see reported yet in this repo, so I'd like to bring it up.

The issue is described here: https://github.com/microsoft/vscode/issues/101136. When alt-clicking in the vscode terminal, if the program doesn't support (something), the screen can fill up with "^[[D" or "^[[C" characters.

Here's a screenshot:
![image](https://user-images.githubusercontent.com/18034092/97321626-77674a00-1845-11eb-9d4d-85c49aa19ad8.png)

In the above vscode issue, @Tyriar linked this PR: https://github.com/xtermjs/xterm.js/pull/896 which seems to have added the alt-click feature. There's apparently no way to disable it, and there's no workaround.

Changing the `"editor.multiCursorModifier"` setting was suggested as a workaround, but it didn't work for me.

Thanks

**Comment by mofux** [comment]
The `editor.multiCursorModifier` vscode setting is editor specific and does not change the terminal alt+click behaviour. 

There currently is no option available to disable the alt+click behaviour. 

I think we would be open to a PR that introduces such a config option (maybe call it `disableAltClickCursorMove` which defaults to `false` to retain backwards compatibility)?

Code pointers:

https://github.com/xtermjs/xterm.js/blob/8e151ee6ef8782f59828ca379a5ad5a0e641abd6/src/browser/services/SelectionService.ts#L670-L686

**Comment by jerch** [comment]
Strong upvote from my side for being able to disable that feature completely, as it is basically broken :smile_cat:.

On a side note: Those tons of nonsense output are cursor movement sequences, that the canonical mode of the TTY cannot handle and instead echoes them back to the user. This will happen for most programs, that are not specially prepared for cursor movement and run the TTY in raw mode (you can test it with `read` in a bash script).

**Comment by mofux** [comment]
I wouldn't call it broken, but rather "unreliable by design" 😬. We are basically using the same technique that is used by `Terminal.app`, which btw. produces the exact same problem:

<img width="842" alt="Screenshot 2020-10-27 at 20 09 35" src="https://user-images.githubusercontent.com/2785983/97349901-56781680-1890-11eb-854b-39f4a5bd7374.png">

**Comment by jerch** [comment]
The problem with this functionality is that is assumes a specific operation mode across the TTY and the slave program, it only works correctly if the line editor of the TTY is switched off (non-canonical) and the slave program has its own terminal view representation which allows cursor movements. This is typically the case for editors, anything else is likely to fail (vast majority of programs).

Shells are awkward in that case - some use canonical mode (ash/dash), shells like bash or zsh operate in raw mode and would work for right/left cursor movements, but not up/down as the history is typically linked against up/down.

REPLs in general share that problem, most very simplistic REPLs use canonical mode and would fail as well, more advanced ones do the full raw mode/own view repr thing and again link their history against up/down.

Last but not least there are curses programs, that might do whatever they want with cursor key sequences.

To summarize this:
The cursor key functionality is not predetermined, it is solely dependent on the slave program, what to make of it. Furthermore there is no way to tell the terminal side that a program is using cursor keys for text cursor movements. If we want to preserve the Alt-Click feature in a reliable way, imho only chance is to come up with a terminal sequence, that allows to switch on/off that feature. Then a program can decide on its own, whether to make use of this terminal feature. But thats kinda redundant, because the mouse terminal API already is capable to do this (simply activate mouse support from your program, listen for Alt-Click and move the cursor within your program accordingly).

**TL;DR:**
This feature is a false promise (to make it blunt: should not exist at all). There is no way to get it working reliably (even the most advanced heuristic will keep failing as long the slave program does not announce the support in a certain way).

**Comment by Tyriar** [comment]
👍 to adding a setting to disable this, that's needed to fix https://github.com/microsoft/vscode/issues/101136 too so VS Code can disable it when `"editor.multiCursorModifier": "ctrlCmd"` is used. I'm not sure it's worth a setting in VS Code to explicitly disable it even when `editor.multiCursorModifier` is the default I don't think alt click does anything, so users would either be pleasantly surprised to discover the feature or confused and not do it again.

FWIW I don't use alt+click so it doesn't impact me much, but I know some do and the idea came from established terminal(s).

And definitely agree it's "unreliable by design" 😄

**Comment by Ranguna** [comment]
Any updates on this ?
I think this has been reported way back in June #101136 and I couldn't find any option to disable/remove/fix this behavior :(
Vscode gets reeeally slow when this happens, I have to scroll away from the ^[[D or else I can't code anything.

**Comment by mofux** [comment]
As stated above, we (the maintainers) would be happy to accept a PR from the community that makes this behaviour configurable.

**Comment by awmottaz** [comment]
I will take a crack at it
